File: cracking_the_coding_interview_go/chapter01/1.2/check_permutation.go
package checkpermutation

import (
	"sort"
)

// Helper function for sorting string
func sortString(s string) string {
	r := []rune(s)
	sort.Slice(r, func(i, j int) bool {
		return r[i] < r[j]
	})
	return string(r)
}

// CheckPermutation -
// First of all, find out these things: Should the comparison be case sensitive?
// Example is God a permutation of dog?
//
// Second: Is whitespaces significant?
// Example "god      " is different from "dog"
func CheckPermutation(word, permutation string) bool {
	// check so the length matches
	if len(word) != len(permutation) {
		return false
	}

	return sortString(word) == sortString(permutation)
}

// CheckPermutationEfficient -
// This function is better if efficient is important, maybe not so readable as the first one
// regarding assumption, always check about the size of the character set, here we assume it is ASCII
func CheckPermutationEfficient(word, permutation string) bool {
	if len(word) != len(permutation) {
		return false
	}

	letters := new([128]int) // Assumption ASCII letters
	runes := []rune(word)

	// count number of each rune in word
	for _, r := range runes {
		letters[r]++
	}

	for i := 0; i < len(permutation); i++ {
		c := permutation[i]
		letters[c]--
		if letters[c] < 0 {
			return false
		}
	}

	return true
}

File: cracking_the_coding_interview_rust/libs/chapter01/src/check_permutation.rs
use itertools::Itertools;

/// First of all, find out these things: Should the comparison be case sensitive?
/// Example is God a permutation of dog?
///
/// Second: Is whitespaces significant?
/// Example "god     " is different from "dog"
pub fn check_permutation(word: &str, permutation: &str) -> bool {
    if word.len() != permutation.len() {
        return false;
    }

    word.chars().sorted().rev().collect::<String>()
        == permutation.chars().sorted().rev().collect::<String>()
}

/// This function is better if efficient is important, maybe not so readable as the first function
/// regarding the assumption, always check about the size of the character set, here we assuming it's ASCII
pub fn check_permutation_efficient(word: &str, permutation: &str) -> bool {
    if word.len() != permutation.len() {
        return false;
    }

    let mut letters: [u8; 128] = [0; 128];

    word.chars().for_each(|ch| {
        letters[ch as usize] += 1;
    });

    (0..permutation.len()).for_each(|i| {
        let c = permutation.chars().nth(i).unwrap();
        letters[c as usize] -= 1;
    });

    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_permutation() {
        assert!(check_permutation("cat", "tac"));
        assert!(!check_permutation("cat", "bar"));
    }

    #[test]
    fn test_permutation_efficient() {
        assert!(check_permutation_efficient("hello", "olleh"));
    }
}
